### PR TITLE
Fixes moving after weight norm application

### PR DESCRIPTION
**Pull request by mruberry**
This PR updates how RNNs handle their "flat weights." In particular, it allows for only some flat weights to be "materialized" when apply is called, and it updates the flattening behavior to only apply if all flat weights are (1) materialized, (2) share a dtype and (3) are acceptable to cuDNN.

One test is modified and another created to test these changes. One practical effect of this change is that weight norm can be successfully applied to a module BEFORE that module is moved to an accelerator. Previously doing so would throw an error.

### CODE REVIEW

**Comment by ngimel** [comment]
fixes #32346

**Review by ngimel — APPROVED** [review]
LGTM

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 15a8a85a:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=15a8a85a17c582f1065e6c792feee5ea30c68e98).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4328684) pytorch_linux_xenial_py3_clang5_asan_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90082002) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90082002))
<details>
<summary>
<code>Jan 23 23:35:11 caused by: Connection refused (os error 111)</code>
</summary>

```
Jan 23 23:35:11 +++ eval 'extract_trap_cmd ' 
Jan 23 23:35:11 ++++ extract_trap_cmd 
Jan 23 23:35:11 ++++ printf '%s\n' '' 
Jan 23 23:35:11 +++ printf '%s\n' cleanup 
Jan 23 23:35:11 ++ trap -- ' 
Jan 23 23:35:11 cleanup' EXIT 
Jan 23 23:35:11 ++ which sccache 
Jan 23 23:35:11 ++ sccache --stop-server 
Jan 23 23:35:11 Stopping sccache server... 
Jan 23 23:35:11 error: couldn't connect to server 
Jan 23 23:35:11 caused by: Connection refused (os error 111) 
Jan 23 23:35:11 ++ true 
Jan 23 23:35:11 ++ rm /var/lib/jenkins/sccache_error.log 
Jan 23 23:35:11 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Jan 23 23:35:11 ++ true 
Jan 23 23:35:11 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Jan 23 23:35:11 ++ SCCACHE_IDLE_TIMEOUT=1200 
Jan 23 23:35:11 ++ RUST_LOG=sccache::server=error 
Jan 23 23:35:11 ++ sccache --start-server 
Jan 23 23:35:11 Starting sccache server... 
Jan 23 23:35:11 ++ sccache --zero-stats 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -124,7 +124,7 @@ def flatten_parameters(self):
         if len(unique_data_ptrs) != len(self._flat_weights):
             return
 
-        with torch.cuda.device_of(any_param):
+        with torch.cuda.device_of(first_fw):
             import torch.backends.cudnn.rnn as rnn
 
             # NB: This is a temporary hack while we still don't have Tensor
```
Comment (ngimel) [inline torch/nn/modules/rnn.py]: Can you remove this comment while you are at it? It is no longer accurate.

**Comment by ezyang** [comment]
this broke lint on master

**Comment by mruberry** [comment]
> this broke lint on master

Sorry about that. Thanks for reverting!

**Pull request closed without merging**